src/poly/mod.rs

```diff
@@ -13,28 +13,30 @@ pub mod mielim;
 pub mod reduce;
 pub mod simplify;
 
-pub trait ToExpr<F, V> {
-    fn expr(&self) -> Expr<F, V>;
+pub trait ToExpr<F, V, M> {
+    fn expr(&self) -> Expr<F, V, M>;
 }
 
 pub trait ToField<F> {
     fn field(&self) -> F;
 }
 
 #[derive(Clone)]
-pub enum Expr<F, V> {
+pub enum Expr<F, V, M> {
     Const(F),
-    Sum(Vec<Expr<F, V>>),
-    Mul(Vec<Expr<F, V>>),
-    Neg(Box<Expr<F, V>>),
-    Pow(Box<Expr<F, V>>, u32),
+    Sum(Vec<Expr<F, V, M>>),
+    Mul(Vec<Expr<F, V, M>>),
+    Neg(Box<Expr<F, V, M>>),
+    Pow(Box<Expr<F, V, M>>, u32),
     Query(V),
     Halo2Expr(Expression<F>),
 
-    MI(Box<Expr<F, V>>), //  Multiplicative inverse, but MI(0) = 0
+    MI(Box<Expr<F, V, M>>), //  Multiplicative inverse, but MI(0) = 0
+
+    Metadata(M),
```

Comment:
@rutefig what is this for?

M should be added to all the fields, like `Sum(Vec<Expr<F, V, M>>, M)`

Reply:
So each of the nodes can have metainfo, so a additional field of type M is added to all of them

Reply:
Yes I was divided into doing that or doing a separate variant because it simplifies the integration, but then we needed to ensure each node had this field. I was just moving forward with one type of solution knowing it could be not the one you had imagined. So I'm going to rewind and move with the one you said

Reply:
Yes, is that Metadata(M) I don't think has an use. I know it is a lot of repetitibve work, but we are going to use it in many places.

Reply:
Yeah it's okay, did the change :)